models/repo_list.go

```diff
@@ -110,8 +110,28 @@ type SearchRepoOptions struct {
 	// maximum: setting.ExplorePagingNum
 	// in: query
 	PageSize int `json:"limit"` // Can be smaller than or equal to setting.ExplorePagingNum
+	// Type of repository to search, related to owner
+	//
+	// in: query
+	SearchMode SearchMode `json:"type"`
 }
 
+// SearchMode is repository filtering mode identifier
+type SearchMode string
+
+const (
+	// SearchModeAny any mode (default)
+	SearchModeAny SearchMode = ""
+	// SearchModeFork fork mode
+	SearchModeFork = "FORK"
+	// SearchModeMirror mirror mode
+	SearchModeMirror = "MIRROR"
+	// SearchModeSource source mode
+	SearchModeSource = "SOURCE"
+	// SearchModeCollaborative collaborative mode
+	SearchModeCollaborative = "COLLABORATIVE"
```

Comment:
I think this approach has some disadvantages:
- What if I want to search for all non-mirror repos (both forks and non-forks)? Or all collaborative forks that are forks? With the current set of search modes, I can't.
- If I use `SearchModeCollaborative`, I must set `Collaborate=true` otherwise I won't get any results. If I want collaborative repos, I shouldn't have to say so twice. Ideally, different parameters would not "interfere" with each other in this way.

I propose the following:
```
// None -> include collaborative AND non-collaborative
// True -> include just collaborative
// False -> incude just non-collaborative
Collaborate util.OptionalBool
// None -> include forks AND non-forks
// True -> include just forks
// False -> include just non-forks
Fork util.OptionalBool
// None -> include mirrors AND non-mirrors
// True -> include just mirrors
// False -> include just non-mirrors
Mirror util.OptionalBool
```

This allows the caller to specify any combination of conditions, and prevents weird interactions between different parameters.

Reply:
Your proposition is good and I want to implement this kind of repository search in the future (eg v2 api). But this PR has different purpose and `SearchMode` has different meaning. It should mimic behaviour and rules used in repository search on dashboard page.

Reply:
@ethantkoenig I think I've over complicated it and you can be right... There is simplier solution to achieve what I wanted. I'll be thinking about it.